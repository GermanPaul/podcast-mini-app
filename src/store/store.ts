import { configureStore } from '@reduxjs/toolkit'
import podcastFeedReducer from '@/store/slices/podcast'

export const store = configureStore({
  reducer: {
    podcastFeed: podcastFeedReducer
  }
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
