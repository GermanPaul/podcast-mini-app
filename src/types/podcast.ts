export interface TopPodcasts {
  feed: Feed
}

export interface Feed {
  author: Author
  entry: Entry[]
  updated: Updated
  rights: FeedRights
  title: FeedTitle
  icon: Icon
  link: FeedLink[]
  id: FeedId
}

export interface Author {
  name: Name
  uri: Uri
}

export interface Name {
  label: string
}

export interface Uri {
  label: string
}

export interface Entry {
  'im:name': ImName
  'im:image': ImImage[]
  summary: Summary
  'im:price': ImPrice
  'im:contentType': ImContentType
  rights: Rights
  title: Title
  link: Link
  id: Id
  'im:artist': ImArtist
  category: Category
  'im:releaseDate': ImReleaseDate
}

export interface ImName {
  label: string
}

export interface ImImage {
  label: string
  attributes: ImImageAttributes
}

export interface ImImageAttributes {
  height: string
}

export interface Summary {
  label: string
}

export interface ImPrice {
  label: string
  attributes: PriceAttributes
}

export interface PriceAttributes {
  amount: string
  currency: string
}

export interface ImContentType {
  attributes: ContentTypeAttributes
}

export interface ContentTypeAttributes {
  term: string
  label: string
}

export interface Rights {
  label: string
}

export interface Title {
  label: string
}

export interface Link {
  attributes: LinkAttributes
}

export interface LinkAttributes {
  rel: string
  type: string
  href: string
}

export interface Id {
  label: string
  attributes: IdAttributes
}

export interface IdAttributes {
  'im:id': string
}

export interface ImArtist {
  label: string
  attributes: ArtistAttributes
}

export interface ArtistAttributes {
  href: string
}

export interface Category {
  attributes: CategoryAttributes
}

export interface CategoryAttributes {
  'im:id': string
  term: string
  scheme: string
  label: string
}

export interface ImReleaseDate {
  label: string
  attributes: ReleaseDateAttributes
}

export interface ReleaseDateAttributes {
  label: string
}

export interface Updated {
  label: string
}

export interface FeedRights {
  label: string
}

export interface FeedTitle {
  label: string
}

export interface Icon {
  label: string
}

export interface FeedLink {
  attributes: Attributes9
}

export interface Attributes9 {
  rel: string
  type?: string
  href: string
}

export interface FeedId {
  label: string
}
