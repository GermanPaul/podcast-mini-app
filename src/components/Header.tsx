import React from 'react'
import { useSelector } from 'react-redux'
import { RootState } from '@/store/store'
import { useI18n } from '@/hooks/useI18n'

import { Link } from 'react-router-dom'

const Header = () => {
  const { t } = useI18n();
  const isLoading = useSelector((state: RootState) => state.podcastFeed.isLoading)

  return <nav className='p-3 border-b border-b-gray-200 flex justify-between'>
    <Link to='/' className='text-2xl font-bold text-cyan-600'>
      {t('HEADER_TITLE')}
    </Link>
    {isLoading && <div className='flex items-center'>
      <svg className='animate-spin h-5 w-5' viewBox='0 0 24 24'>
        <circle 
          className='opacity-25 bg-white' 
          cx='12' 
          cy='12' 
          r='10' 
          stroke='currentColor' 
          fill='white' 
          strokeWidth='4'>
        </circle>
        <path 
          className='opacity-75 text-cyan-600' 
          fill='currentColor' 
          d='M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z'>
        </path>
      </svg>
      <span className='mx-1'>{t('HEADER_LOADING')}...</span>
    </div>}
  </nav>
}

export default Header
