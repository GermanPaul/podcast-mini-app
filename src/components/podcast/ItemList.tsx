import React from 'react'
import { PodcastDetailsItem } from '@/types/podcastDetailsApi'
import { msToTime } from '@/utils'
import { useI18n } from '@/hooks/useI18n'

import { Link } from 'react-router-dom'

interface ItemListProps {
  items: PodcastDetailsItem[]
  podcastId: string
}

const ItemList = ({ items, podcastId }: ItemListProps) => {
  const { t } = useI18n()

  return <div className='px-4'>
    <div className='shadow-md p-3 mb-2 text-lg font-bold'>{t('ITEM_EPISODES')}: {items.length}</div>
    <div className='shadow-md p-3'>
      <table className='w-full'>
        <thead className='border-b-2'>
          <tr>
            <th className='w-2/3 text-left'>{t('ITEM_TITLE')}</th>
            <th className='w-1/6 text-left'>{t('ITEM_DATE')}</th>
            <th className='w-1/6 text-left'>{t('ITEM_DURATION')}</th>
          </tr>
        </thead>
        <tbody className='[&>*:nth-child(odd)]:bg-gray-100'>
          {
            items.map((item, index, array) => {
              const date = new Date(item.date)
              const duration = isNaN(item.duration as any) ?
                item.duration :
                msToTime(parseInt(item.duration))

              return <tr className='h-8' key={`row-${index + 1}`}>
                <td><Link className='text-sky-700' to={`/podcast/${podcastId}/episode/${item.id}`}>{item.title}</Link></td>
                <td>{date.toLocaleDateString('en')}</td>
                <td>{duration}</td>
              </tr>
            }
            )
          }
        </tbody>
      </table>
    </div>
  </div>
}

export default ItemList
