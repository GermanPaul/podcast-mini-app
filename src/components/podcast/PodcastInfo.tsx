import React from 'react'
import { PodcastDetails } from '@/types/podcastDetailsApi'
import { useI18n } from '@/hooks/useI18n'

interface PodcastInfoProps {
  podcast: PodcastDetails
}

const PodcastInfo = ({ podcast }: PodcastInfoProps) => {
  const { t } = useI18n()

  return <div className='p-4 shadow-md'>
    <img className='mx-auto mb-4 rounded-md w-72 h-72' src={podcast.image} alt={podcast.title} loading='lazy' />
    <div className='my-3 mx-4'>
      <div className='font-bold'>{podcast.title}</div>
      <div className='italic'>{t('PODCAST_BY')} {podcast.author}</div>
    </div>
    <div className='my-3 mx-4'>
      <div className='font-bold'>{t('PODCAST_DESCRIPTION')}</div>
      <div dangerouslySetInnerHTML={{__html: podcast.description}}></div>
    </div>
  </div>
}

export default PodcastInfo
