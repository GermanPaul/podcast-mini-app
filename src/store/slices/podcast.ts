import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Feed } from '@/types/podcast'

interface PodcastFeedState {
  isLoading: boolean
}

const initialState: PodcastFeedState = {
  isLoading: false
}

const podcastFeedSlice = createSlice({
  name: 'podcastFeed',
  initialState,
  reducers: {
    setIsLoading: (state, action: PayloadAction<boolean>) => {
      state = { ...state, isLoading: action.payload}
    }
  }
})

export const { setIsLoading } = podcastFeedSlice.actions

export default podcastFeedSlice.reducer
