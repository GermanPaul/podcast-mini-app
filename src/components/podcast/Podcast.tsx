import React from 'react'
import { useParams } from 'react-router-dom'
import { useFetchPodcastDetails } from '@/hooks/api/useFetchPodcastDetails'
import { PodcastDetails, PodcastDetailsItem } from '@/types/podcastDetailsApi'

import PodcastInfo from '@/components/podcast/PodcastInfo'
import ItemList from '@/components/podcast/ItemList'
import Skeleton from 'react-loading-skeleton'

const Podcast = () => {
  const { podcastId } = useParams<{ podcastId: string }>()
  const { isLoading, data: podcastDetails, isError } = useFetchPodcastDetails(podcastId as string)

  if (isError) {
    return <div className='p-3 text-lg'>
      Unable to fetch Podcast details.
    </div>
  }

  if (isLoading) {
    return <div className='max-w-7xl mx-auto flex p-2'>
      <div className='basis-2/6 m-2 shrink-0'>
        <Skeleton className='animate-pulse' enableAnimation={true} height='30rem' />
      </div>
      <div className='basis-4/6 m-2'>
        <Skeleton className='animate-pulse' height='30rem' />
      </div>
    </div>
  }

  return <div>
    <div className='max-w-7xl mx-auto flex p-2'>
      <div className='basis-2/6 m-2 shrink-0'>
        {
          !!podcastDetails && (<PodcastInfo podcast={podcastDetails as PodcastDetails} />)
        }
      </div>
      <div className='basis-4/6 m-2'>
        {
          !!podcastDetails && <ItemList items={podcastDetails?.items as PodcastDetailsItem[]} podcastId={podcastId as string} />
        }
      </div>
    </div>
  </div>
}

export default Podcast
