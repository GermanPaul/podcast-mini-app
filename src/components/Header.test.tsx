import { act, queryByText, render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import { Provider } from 'react-redux';
import { store } from '@/store/store'
import { setIsLoading } from '@/store/slices/podcast'
import { IntlProvider } from 'react-intl'
import { BrowserRouter } from 'react-router-dom'

import Header from './Header';

describe('<Header />', () => {
  it('render the component', () => {
    store.dispatch(setIsLoading(false))

    const { baseElement } = render(<Provider store={store}>
      <IntlProvider
        locale='en' 
        messages={{ HEADER_LOADING: 'Loading', HEADER_TITLE: 'Podcaster', }}
      >
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </IntlProvider>
    </Provider>)

    expect(baseElement).toMatchSnapshot()
  })

  it('should show loading state', async () => {
    store.dispatch(setIsLoading(true))

    render(<Provider store={store}>
      <IntlProvider
        locale='en' 
        messages={{ HEADER_LOADING: 'Loading', HEADER_TITLE: 'Podcaster', }}
      >
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </IntlProvider>
    </Provider>)

    expect(screen.queryByText('Loading...')).toBeInTheDocument()

    act(() => {
      store.dispatch(setIsLoading(false))
    })

    await waitFor(() => expect(screen.queryByText('Loading...')).toBeNull())
  })

  it('should display a home link', async () => {
    store.dispatch(setIsLoading(true))

    render(<Provider store={store}>
      <IntlProvider
        locale='en' 
        messages={{ HEADER_LOADING: 'Loading', HEADER_TITLE: 'Podcaster', }}
      >
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </IntlProvider>
    </Provider>)

    const headerLink = screen.getByRole('link')
    
    expect(headerLink).toHaveProperty('href', '/')
    expect(headerLink).toHaveTextContent('Podcaster')
  })
})
