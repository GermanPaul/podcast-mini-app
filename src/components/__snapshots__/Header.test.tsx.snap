// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header /> > render the component 1`] = `
<body>
  <div>
    <nav
      class="p-3 border-b border-b-gray-200 flex justify-between"
    >
      <a
        class="text-2xl font-bold text-cyan-600"
        href="/"
      >
        Podcaster
      </a>
    </nav>
  </div>
</body>
`;
